### ISSUE TITLE
Please add information for devices with WebViewStub and Magisk to the Wiki page

### ISSUE THREAD

**Issue by ubergeek77** [bromite/bromite#243]
Currently, the [Wiki page on installing Bromite System WebView](https://github.com/bromite/bromite/wiki/Installing-SystemWebView) assumes the target device has a `/system/app/webview` folder.

However, some devices, such as my Google Pixel 3 XL on Android 9.0, do not have this folder, and only have `/system/app/WebViewStub.`

Additionally, due to Magisk being the only root method for this device, I am unable to use the suggested procedure using "/system/app" mover.

I did use a Magisk module, App Systemizer, to simulate this step. This gave me the option to move the app to /system/app or /system/app-priv using Magisk's "systemless" approach.

Unfortunately, even after installing the app again over the system version, I'm still unable to select Bromite System WebView as a WebView provider.

I've tried a few different things manually via TWRP to get this installed, but nothing seems to work. I tried using a Magisk module provided by NanoDroid, but apparently it is incompatible with the way my Phone's TWRP mounts system directories, so that installation did not work for me.

If the Wiki page could be updated to account for devices where the only root option is Magisk, and for devices that have `/system/app/WebViewStub` instead of `/system/app/webview`, it would be greatly appreciated. I really want to get this installed.

As a side note to everyone reading this, PLEASE be careful when mucking about with the WebView apps. I caused a boot loop and was unable to boot my phone while doing this. I was able to use adb during the boot process, and logcat very clearly showed that the cause of the boot loop was an error with the WebView app package. With this being a privacy focused app, some of us here may also be using Signal. I'd like to remind everyone that Signal Private Messenger employs some VERY strict encryption/keystone techniques, and even if you have a TWRP backup, you may find yourself unable to open Signal after the restore. I lost all of my messages like this, though I am still exploring decryption options. Be sure to make a backup from within Signal if you care about this, and backup the SignalSecret keystore file under /data/misc/keystore/user_0.

Please learn by my example and don't repeat the mistake I made by losing my Signal messages.

**Comment by csagan5** [comment]
@ubergeek77 the whole SystemWebView is provided "as is", there is nothing officially supported (so not Magisk, nor any specific Android version). You need some skills to install it and that wiki page tries to give the easiest approach.

As I understand that approach does not work with Android Pie 9.0; I am going to write that on the wiki page but there is nothing else to write there as you do not seem to describe how to make it work on Android 9.0.

**Comment by ubergeek77** [comment]
Thank you for adding that info to the wiki page. The included video was captured from a Pie device (since you can see the USB Debugging "P" icon), so I assumed I was in the clear in terms of installing this. It's weird that adding it as a system app via Magisk didn't work. Perhaps this incompatibility is a result of the way Magisk remounts /system, so module installation is not possible. Hopefully someone with enough interest can come along and figure this out.

As someone with much more knowledge of WebView than I, I feel obliged to ask you:

If I wanted to do this manually via TWRP, would I need to replace the /system/app/WebViewStub folder? Or do I simply need to add Bromite System WebView to the /system/app or /system/app-priv folders? Does it matter if the app goes in app or app-priv?

I'm trying to wrap my head around how Android determines a system app is a "valid" WebView provider. I'd be happy to test an alternative installation procedure so that it can be added to the Wiki, if you can think of one.

**Comment by csagan5** [comment]
> If I wanted to do this manually via TWRP, would I need to replace the /system/app/WebViewStub folder? Or do I simply need to add Bromite System WebView to the /system/app or /system/app-priv folders? Does it matter if the app goes in app or app-priv?

I have no idea, never installed it on Android Pie. You should ask on Reddit or XDA, there is no user support here on Github but simply issue tracking for bugs and features.

> I'm trying to wrap my head around how Android determines a system app is a "valid" WebView provider.

The SystemWebView installation wiki page has some explanation, but this is not something specific to Bromite so internet is your resource for knowing more.



> I'd be happy to test an alternative installation procedure so that it can be added to the Wiki, if you can think of one.

There is already an installation procedure for Android Pie using system app mover; if you cannot use that then it means that your ROM does not support this. You can check with instruction on the wiki page about which webviews your ROM supports.

**Comment by ahmadsadidut** [comment]
In my rooted Oppo a37f phone, lolypop, lucky patcher instaled, can i installing bromite webview ? Already have system/app/WebViewGoogle/WebViewGoogle.apk Or need some apk modification...

**Comment by Bluscream** [comment]
Same ^

**Comment by lss4** [comment]
> Thank you for adding that info to the wiki page. The included video was captured from a Pie device (since you can see the USB Debugging "P" icon), so I assumed I was in the clear in terms of installing this. It's weird that adding it as a system app via Magisk didn't work. Perhaps this incompatibility is a result of the way Magisk remounts /system, so module installation is not possible. Hopefully someone with enough interest can come along and figure this out.
> 
> As someone with much more knowledge of WebView than I, I feel obliged to ask you:
> 
> If I wanted to do this manually via TWRP, would I need to replace the /system/app/WebViewStub folder? Or do I simply need to add Bromite System WebView to the /system/app or /system/app-priv folders? Does it matter if the app goes in app or app-priv?
> 
> I'm trying to wrap my head around how Android determines a system app is a "valid" WebView provider. I'd be happy to test an alternative installation procedure so that it can be added to the Wiki, if you can think of one.

Apparently the wiki mentioned this. You can check your ROM's framework-res.apk, decode the /res/xml/config_webview_packages.xml and see if com.android.webview is allowed to be used as WebView.

The thing is, some stock Pie ROMs have a stricter requirement. They do not allow AOSP webview (com.android.webview), only Google WebView, and it MUST be a system app. You cannot use a sideloaded Chrome as WebView on such ROMs. I have a Xperia XZ Premium and its stock Pie ROM (47.2.A.x.x) has such limitations.

This limitation presents a problem not only with Bromite but also with OpenGApps lesser than stock (on which Chrome was bundled), like Micro/Nano/Pico, which can result in a crippled system with no usable WebView (see [this issue](https://github.com/opengapps/opengapps/issues/724) for details).

One might be able to work around this limitation by spoofing it as com.google.android.webview, if the xml doesn't require a signature for it (which is my case). This can help getting it usable on some stock ROMs. Still, Bromite SystemWebView is mostly helpful for custom ROM builders (such as LineageOS), that one can use this in their own ROM builds in place of AOSP WebView.

**Comment by csagan5** [comment]
> One might be able to work around this limitation by spoofing it as com.google.android.webview, if the xml doesn't require a signature for it (which is my case).

I doubt that a Google Webview is allowed without signature; even if there is no signature there, it is guarded against replacement by other means. See the issue tracker links on the wiki page.

**Comment by lss4** [comment]
> > One might be able to work around this limitation by spoofing it as com.google.android.webview, if the xml doesn't require a signature for it (which is my case).
> 
> I doubt that a Google Webview is allowed without signature; even if there is no signature there, it is guarded against replacement by other means. See the issue tracker links on the wiki page.

It seems [this thread](https://groups.google.com/a/chromium.org/forum/#!topic/android-webview-dev/nZd-bNQfu80) has the answer.

> If this APK is installed by default (in the system image), then a signature is not mandatory. You can look at config_webview_packages.xml as an example ("com.android.webview" is installed by default and has no signature). You should use 'availablebyDefault="true"' for this entry.
> 
> In terms of security, the signing key checks are already performed by the system upon APK installation (you can't update an APK over an existing one unless signing keys match). The signature is only necessary for non-preinstalled providers (because the platform wouldn't already perform that security check).

Apparently, it turned out the reason why com.android.chrome didn't show up as a valid provider even after installing from Play Store (on my stock ROM which doesn't already have it bundled) was also because of the `availableByDefault="true"` attribute. That attribute seems to specify the need of it being a system app (so that you don't need to actually provide a signature since the system already checked it during upgrade).

I just tried installing Chrome Beta from Play Store (which doesn't have the `availableByDefault` attribute but has a hardcoded signature) and it did show up in the valid provider list. However, sideloading a package (like those from APKPure) would not make it show up since the packages there have a different signature, but taking out the base.apk from your system, uninstall the installed one, then sideloading that base.apk back would still make it show up as the base.apk has a matching signature.

Still, it doesn't answer what if we put our own "com.google.android.webview" in place of the real Google WebView, either directly into the system partition, or systemlessly via Magisk/NanoDroid, under such situation (with `availableByDefault="true"` but no signature).

**Comment by csagan5** [comment]
> Still, it doesn't answer what if we put our own "com.google.android.webview" in place of the real Google WebView, either directly into the system partition, or systemlessly via Magisk/NanoDroid, under such situation (with `availableByDefault="true"` but no signature).

It has been tried many times, always with negative results. You can try it yourself by editing the APK package name and re-signing it with a different key.

@lss4 you might have already individuated useful information to edit the wiki page, do you feel like incorporating some information there?

**Comment by lss4** [comment]
I've been tinkering with config_webview_packages.xml in my stock ROM's framework-res.apk, and from what I've tested:

1. You need at least one WebView provider with the `availableByDefault="true"` attribute, or the system will bootloop with this error in the logcat:

> 07-13 23:38:39.719  1401  1401 E System  : Caused by: android.util.AndroidRuntimeException: There must be at least one WebView package that is available by default
> 07-13 23:38:39.719  1401  1401 E System  : 	at com.android.server.webkit.SystemImpl.<init>(SystemImpl.java:140)
> 07-13 23:38:39.719  1401  1401 E System  : 	at com.android.server.webkit.SystemImpl.<init>(SystemImpl.java:55)
> 07-13 23:38:39.719  1401  1401 E System  : 	at com.android.server.webkit.SystemImpl$LazyHolder.<clinit>(SystemImpl.java:70)
> 07-13 23:38:39.719  1401  1401 E System  : 	... 11 more

Actually, more info can be found in [SystemImpl.java](https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/webkit/SystemImpl.java#77), in the class' constructor.

2. A provider with `isFallback="true"` attribute must also have `availableByDefault="true"`, and you can only have one fallback provider (according to SystemImpl.java). The fallback provider would be automatically frozen by the system if you have another qualified provider present.

3. If a provider has `availableByDefault="true"` attribute, it does not require a `signature` block to be qualified, but it must be a system app. Such provider can still be qualified, if replaced with another one with the same package name but with different signature. As a result, adding a com.android.webview entry with `availableByDefault="true"` in the XML would make the Bromite SystemWebView NanoDroid module usable.

4. A provider with neither `availableByDefault="true"` nor a `signature` block could never qualify. If you don't want to use NanoDroid but want to sideload Bromite SystemWebView instead, you'll need to explicitly specify its signature, which can be extracted from Bromite SystemWebView's apk with this command (from [the aforementioned thread](https://groups.google.com/a/chromium.org/forum/#!topic/android-webview-dev/nZd-bNQfu80)):
`openssl pkcs7 -in BROMITEK.RSA -inform DER -print_certs`
And put the signature information you got into a `signature` block within the com.android.webview entry.

**Comment by ubergeek77** [comment]
So we know that we need to modify values in that .xml, which is inside of framework-res.apk.

As far as the stock Pixel ROMs go, that APK is impossible to decompile and then recompile using something like apktool (apktool's author refers to this as "magic apks.). And from experience, as a rule of thumb, unless you're using an AOSP-based ROM, you won't be decompiling that .apk.

That being said, is there any other way to replace that .xml, perhaps at runtime? I've tried messing around with .apk overlays with no luck. Hopefully someone with more experience will know what to do here.

Can something like Substratum be of use here? I know it's intended for theming, but with the xml that needs to be modified being in framework-res.apk, maybe it can be overridden like that somehow.

**Comment by lss4** [comment]
> So we know that we need to modify values in that .xml, which is inside of framework-res.apk.
> 
> As far as the stock Pixel ROMs go, that APK is impossible to decompile and then recompile using something like apktool (apktool's author refers to this as "magic apks"). And from experience, as a rule of thumb, unless you're using an AOSP-based ROM, you won't be decompiling that .apk - this goes for the vast majority of stock ROMs for pretty much every device.
> 
> That being said, is there any other way to replace that .xml, perhaps at runtime? I've tried messing around with .apk overlays with no luck. Hopefully someone with more experience will know what to do here.
> 
> Can something like Substratum be of use here? I know it's intended for theming, but with the xml that needs to be modified being in framework-res.apk, maybe it can be overridden like that somehow.

Yeah, if you replace the original framework-res.apk with the recompiled one directly, you'll get a bootloop. This [old post](https://forum.xda-developers.com/showpost.php?p=30154154&postcount=5) turned out to be the solution to it. After recompiling, put the META-INF folder and AndroidManifest.XML file from the original framework-res.apk back into it, and you'll get a working framework-res.apk.

If your stock ROM's odexed, you might need to deodex the framework before doing any work.

As for Substratum... depending on whether it can be used to override resources in framework-res.apk. This ([Privset](https://forum.xda-developers.com/android/apps-games/app-t3694602)) might be of interest as it could modify certain values inside framework-res.apk.

**Comment by csagan5** [comment]
> if your system ships with com.android.chrome or com.google.android.webview, instead of com.android.webview, you need to build Bromite SystemWebView with these package names to properly replace them

Have you tested renaming&re-signing a Bromite SystemWebView to validate the above statement?

**Comment by lss4** [comment]
> > if your system ships with com.android.chrome or com.google.android.webview, instead of com.android.webview, you need to build Bromite SystemWebView with these package names to properly replace them
> 
> Have you tested renaming&re-signing a Bromite SystemWebView to validate the above statement?

CONFIRMED: I renamed Bromite SystemWebView's package name to com.android.chrome and put it into the system folder, and it showed up in the WebView provider list, even though it's not the real Chrome.

**Comment by ubergeek77** [comment]
> > > if your system ships with com.android.chrome or com.google.android.webview, instead of com.android.webview, you need to build Bromite SystemWebView with these package names to properly replace them
> > 
> > 
> > Have you tested renaming&re-signing a Bromite SystemWebView to validate the above statement?
> 
> CONFIRMED: I renamed Bromite SystemWebView's package name to com.android.chrome and put it into the system folder, and it showed up in the WebView provider list, even though it's not the real Chrome.
> 
> EDIT: Note that this is only for stock ROMs whose XMLs set the `availableByDefault="true"` attribute instead of a signature block for com.android.chrome and/or com.google.android.webview entries.
> 
> For AOSP ROMs, you can only use com.android.webview, as these ROMs usually support sideloading Chrome Stable or Google WebView, which means signature blocks are provided for these entries.

@lss4 I'd like to test this as well - can you help me out on how you were able to rename the SystemWebView APK? I'm having trouble getting apktool to recompile the decompiled .apk, it just throws a bunch of errors.

**Comment by lss4** [comment]
I used [this utility](https://github.com/sjitech/ApkRename) to rename the APK. You'll need a debug keystore to sign the APK before you can put it into the system/app folder. If you use Android Studio, you might already have a [debug keystore](https://developer.android.com/studio/publish/app-signing#debug-mode) for the purpose, which usually resides in ~/.android/debug.keystore.

You can use any name for the app folder and apk file in the system/app folder. Just make sure the files have the correct permissions or otherwise you risk a bootloop.

I haven't tested whether the renamed APK would work as intended, though. I only confirmed that it did show up on my stock ROM's WebView providers, after renaming it and put it to the system/app folder. It's possible there might be some breakages, if references to the original package name (com.android.webview) were somehow hardcoded in some parts of the bytecode.

If we can confirm this, it might be good to add these package names as flavors, so people can choose which package name to spoof depending on their own stock ROM (whether it be Chrome Stable's or Google WebView's package name).

It's possible this can confuse Google Play Store in regards of package installation status and updates... but from what I've seen, the current Google Play Store appears to be already aware of the cases where a package with the same name but different signature are being installed, and would not attempt to update them.

**Comment by ubergeek77** [comment]
I just tried this, and it was a success! ..... _kinda._

The only line in my config_webview_packages.xml with an available, non-signature'd package name is:

`<webviewprovider description="Chrome Debug" packageName="com.google.android.apps.chrome" />`

So I renamed the Bromite WebView apk to `com.google.android.apps.chrome`. I then used the App Systemizer Magisk Module to convert it to a system app. And it worked! I can see Bromite System WebView as an option in WebView implementation.

Unfortunately, that's it. It doesn't actually work. Any app that tries to use a WebView immediately crashes when the WebView would be launched. To me, four possible things went wrong here:
- The renamed app was not renamed properly
- Android is expecting a different package name for parts of the app at runtime
- The apps I am trying are incompatible with Bromite's patches
- The app was not properly converted to a system app

I have no way of knowing which of these is causing the problem, but barring anything that could potentially put me in a bootloop, I'd be happy to track this down.

**Comment by lss4** [comment]
> I just tried this, and it was a success! ..... _kinda._
> 
> The only line in my config_webview_packages.xml with an available, non-signature'd package name is:
> 
> `<webviewprovider description="Chrome Debug" packageName="com.google.android.apps.chrome" />`
> 
> So I renamed the Bromite WebView apk to `com.google.android.apps.chrome`. I then used the App Systemizer Magisk Module to convert it to a system app. And it worked! I can see Bromite System WebView as an option in WebView implementation.
> 
> Unfortunately, that's it. It doesn't actually work. Any app that tries to use a WebView immediately crashes when the WebView would be launched. To me, four possible things went wrong here:
> 
> * The renamed app was not renamed properly
> * Android is expecting a different package name for parts of the app at runtime
> * The apps I am trying are incompatible with Bromite's patches
> * The app was not properly converted to a system app
> 
> I have no way of knowing which of these is causing the problem, but barring anything that could potentially put me in a bootloop, I'd be happy to track this down.
> 
> Edit: This may be useful; here's a line from my logcat:
> 
> `E AndroidRuntime: java.lang.RuntimeException: Package not found: com.google.android.apps.chrome`

If you have a properly working Bromite then apps shouldn't have any problems with it as far as I can tell.

Renaming does not always work as intended, as in some apps, certain parts of the app may contain hardcoded the reference to the original package name, which breaks if the package is renamed, or if one is trying to build the app under a different package name.

I recall with NanoDroid Bromite SystemWebView module on an AOSP-based ROM, sometimes you need to manually sideload the APK once (to also place a copy as a user app) in order to make apps properly recognize the package. Perhaps you can try that as well.

In conclusion, regardless of whether the renamed package works or not, using `com.android.webview` is definitely not going to work on most of the stock ROMs nowadays, as that's the package name for AOSP WebView which is almost only present (and defined) in AOSP-based custom ROMs such as LineageOS.

**Comment by ubergeek77** [comment]
I really like what you said earlier:

> it might be good to add these package names as flavors, so people can choose which package name to spoof depending on their own stock ROM (whether it be Chrome Stable's or Google WebView's package name).

I 1,000% agree with this. For my current use case, I would like to have the ability to use Bromite WebView _alongside_ the default WebView, just so I can easily switch back to the default one if I ever need to for compatibility reasons. There have been a handful of websites I've encountered recently that freak out when "block ads" is enabled in Bromite, and rather than just disable ad-blocking outright, I'd rather have a way to quickly switch WebView providers through the system settings.

Plus, I'm looking to build and switch to RattlesnakeOS soon, and that builds a Chromium WebView by default. If I'm not mistaken, the package name used by that is the same one used by Bromite. I'd really, _really_ prefer to just have Bromite fully working under a totally different package name, like `org.bromite.webview`, so I can simply register that customized package name to config_webview_packages.xml and finally solve this problem for good.

Realistically speaking, it would be a nightmare for a project of this size and scope to officially maintain multiple package names, so I don't expect this to be an official solution.

However, if @csagan5 can share some insight into what would need to be changed before building Bromite from source to fully change the package name, I'd very much appreciate it. I don't at all mind building this from source, but I'm really not sure how I would go about changing the package name.

Surely it would need to be changed in AndroidManifest.xml, and perhaps the directory names would need to be renamed (i.e. /com/android/webview -> /org/bromite/webview), but I have a sneaking suspicion that only changing those two things aren't going to be sufficient. I'm anticipating some other weird thing that would need to be changed first.

If this actually works, I wouldn't be opposed to creating and sharing a script or Docker container that easily facilitates this change so that others could benefit as well. Perhaps it could be added to the Wiki.
